compiler/pipes/analyze-performance.cpp

```diff
@@ -0,0 +1,636 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2020 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+#include "compiler/pipes/analyze-performance.h"
+
+#include "common/algorithms/string-algorithms.h"
+#include "common/termformat/termformat.h"
+#include "compiler/compiler-core.h"
+#include "compiler/data/src-file.h"
+#include "compiler/inferring/public.h"
+#include "compiler/gentree.h"
+
+namespace {
+
+VertexPtr remove_conv_wrap(VertexPtr vertex) noexcept {
+  if (vk::any_of_equal(vertex->type(), op_conv_int, op_conv_float, op_conv_string, op_conv_array,
+                       op_conv_bool, op_conv_mixed, op_conv_drop_false, op_conv_drop_null)) {
+    return remove_conv_wrap(vertex.as<meta_op_unary>()->expr());
+  }
+  return vertex;
+}
+
+bool is_pure_nary_operation(VertexPtr vertex) noexcept {
+  return vk::any_of_equal(vertex->type(), op_minus, op_plus, op_not,
+                          op_add, op_mul, op_sub, op_div, op_mod, op_pow,
+                          op_and, op_or, op_xor, op_shl, op_shr,
+                          op_log_xor_let, op_log_and_let, op_log_or_let, op_log_and, op_log_or,
+                          op_eq2, op_neq2, op_eq3, op_neq3, op_le, op_ge, op_lt, op_gt,
+                          op_spaceship, op_null_coalesce, op_ternary, op_isset, op_string_build);
+}
+
+template<bool full>
+std::string get_description_for_help_impl(VertexPtr expr);
+
+std::string join_string_build_elements(VertexRange elements) noexcept {
+  return vk::join(elements, ".", get_description_for_help_impl<false>);
+}
+
+template<bool full>
+std::string get_description_for_help_impl(VertexPtr expr) {
```

Comment:
Why full is not just a boolean argument?

Reply:
doesn't matter, just micro optimization